### PR TITLE
Rename DeviceAllocatorRegistrationInfo to a more generic name; Use OrtArenaCfg for arena members; Remove unused OrtMemType; Simplify CreateAllocator interface.

### PR DESCRIPTION
**Pull request by pranavsharma**
**Description**: Rename DeviceAllocatorRegistrationInfo to a more generic name; Use OrtArenaCfg for arena members; Remove unused OrtMemType; Simplify CreateAllocator interface.

**Motivation and Context**
Readability, cleanup, simplification.

### CODE REVIEW

File: onnxruntime/core/framework/allocatormgr.cc [hunk 1]
```diff
@@ -12,21 +12,37 @@
...
+    int initial_chunk_size_bytes = info.arena_cfg.initial_chunk_size_bytes == -1
+                                       ? BFCArena::DEFAULT_INITIAL_CHUNK_SIZE_BYTES
+                                       : info.arena_cfg.initial_chunk_size_bytes;
+    int max_dead_bytes_per_chunk = info.arena_cfg.max_dead_bytes_per_chunk == -1
+                                       ? BFCArena::DEFAULT_MAX_DEAD_BYTES_PER_CHUNK
+                                       : info.arena_cfg.max_dead_bytes_per_chunk;
+    ArenaExtendStrategy arena_extend_str;
+    switch (info.arena_cfg.arena_extend_strategy) {
+      case 1:
+        arena_extend_str = ArenaExtendStrategy::kSameAsRequested;
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocatormgr.cc:29]: nit: slightly opaque to have '1' as the case vs `case static_cast<int>(ArenaExtendStrategy::kSameAsRequested):`

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 2]
```diff
@@ -135,36 +137,37 @@ CUDAExecutionProvider::CUDAExecutionProvider(const CUDAExecutionProviderInfo& in
...
+  AllocatorCreationInfo default_memory_info(
+      {[](OrtDevice::DeviceId device_id) {
          return onnxruntime::make_unique<CUDAAllocator>(device_id, CUDA);
        },
-       cuda_mem_limit_});
+       device_id_,
+       true,
+       {cuda_mem_limit_,
+        arena_extend_strategy == ArenaExtendStrategy::kNextPowerOfTwo ? 0 : 1,
+        -1, -1}});
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:148]: nit: can you just static_cast the value to an int?

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc [hunk 3]
```diff
@@ -87,14 +87,14 @@ MIGraphXExecutionProvider::MIGraphXExecutionProvider(const MIGraphXExecutionProv
...
+      {[](int id) { return onnxruntime::make_unique<HIPAllocator>(id, MIGRAPHX); }, device_id_});
+  allocator_ = CreateAllocator(default_memory_info);
   InsertAllocator(allocator_);
 
-  DeviceAllocatorRegistrationInfo pinned_memory_info(
-      {OrtMemTypeCPUOutput, [](int) { return onnxruntime::make_unique<HIPPinnedAllocator>(0, MIGRAPHX_PINNED); }, std::numeric_limits<size_t>::max()});
-  InsertAllocator(CreateAllocator(pinned_memory_info, device_id_));
+  AllocatorCreationInfo pinned_memory_info(
+      {[](int) { return onnxruntime::make_unique<HIPPinnedAllocator>(0, MIGRAPHX_PINNED); }, std::numeric_limits<size_t>::max()}, device_id_);
+  InsertAllocator(CreateAllocator(pinned_memory_info));
```
Comment (skottmckay) [inline onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:97]: This looks off
```
    {
      [](int) { return onnxruntime::make_unique<HIPPinnedAllocator>(0, MIGRAPHX_PINNED); }, 
      std::numeric_limits<size_t>::max()
    }, 
    device_id_);
```

May be simpler to just call the ctor, and the args seem mismatched

File: onnxruntime/core/session/device_allocator.cc [hunk 4]
```diff
@@ -56,37 +56,32 @@ ORT_API_STATUS_IMPL(OrtApis::CreateAndRegisterAllocator, _Inout_ OrtEnv* env, _I
...
+
     if (arena_cfg) {
       if (arena_cfg->max_mem != -1) max_mem = arena_cfg->max_mem;
-      if (arena_cfg->arena_extend_strategy == 0) {
-        arena_extend_strategy = ArenaExtendStrategy::kNextPowerOfTwo;
-      } else if (arena_cfg->arena_extend_strategy == 1) {
-        arena_extend_strategy = ArenaExtendStrategy::kSameAsRequested;
-      }
+      if (arena_cfg->arena_extend_strategy != 1) arena_extend_strategy = arena_cfg->arena_extend_strategy;
       if (arena_cfg->initial_chunk_size_bytes != -1) initial_chunk_size_bytes = arena_cfg->initial_chunk_size_bytes;
```
Comment (skottmckay) [inline onnxruntime/core/session/device_allocator.cc:70]: This is a bit opaque compared to the previous code

**Review by skottmckay — CHANGES_REQUESTED** [review]
:clock1:

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc (hunk 3 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:97]: This style (mixing of aggregate initialization with constructor call) existed even before this PR. Will fix the arg mismatch. Thx for catching it.

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**